pkg/tracing/options.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import "github.com/spf13/cobra"
+
+// Options defines the set of options supported by Istio's component tracing package.
+type Options struct {
+	// URL of zipkin collector (example: 'http://zipkin:9411/api/v1/spans'). This enables tracing for Mixer itself.
+	ZipkinURL string
+
+	// URL of jaeger HTTP collector (example: 'http://jaeger:14268/api/traces?format=jaeger.thrift'). This enables tracing for Mixer itself.
+	JaegerURL string
+
+	// Whether or not to emit trace spans as log records.
+	LogTraceSpans bool
+}
+
+// NewOptions returns a new set of options, initialized to the defaults
+func NewOptions() *Options {
```

Comment:
Do you expect some initialization to be done here? There doesn't currently seem to be an advantage of just using `Options` directly.

Reply:
Indeed, at the moment there is no benefit in NewOptions rather than just Options{}. I used this pattern to be consistent with pkg/log and to allow for easy upgrade in the future if/when some non-zero default value is necessary/desirable. For example, the http timeout is currently fixed and we may wish to eventually make this an option and  a default of 0 would be unpleasant :-)

Reply:
Totally reasonable to me ... although I can't really speak to the "goishness" of this pattern (I suspect it's fine). Go structs definitely suffer from lack of a good solution for initialization, so I personally like this pattern.

Reply:
This is actually considered a bit of an anti-pattern in go. This isn't doing any arg validation or anything. I'm not sure we want to start adding default constructors all over the place. That being said, I don't know how strict we want to be about this kinda thing.

Reply:
I think the lack of enforced constructors in Go is itself an anti-pattern :-)

The only way I can ensure forward compatibility here with cases where the default is not zero-initialized memory is to put in place NewOptions right now and document it as being the way to get things done.